nestle_core/src/macros/derive.rs

```diff
@@ -144,8 +144,8 @@ fn enum_parts(
     (
         width,
         quote! {
-            let shift = 64 - offset - #self_width;
-            *id |= ((unsafe { *<*const _>::from(self).cast::<#repr>() } as i64 as u64) << shift) as i64;
+            let shift = 64 - #self_width;
+            *id += ((unsafe { *<*const _>::from(self).cast::<#repr>() } as u64) << shift >> offset) as i64;
```

Comment:
I’m not sure this will work, but I can’t see what’s different here other than the cast to i64 before the cast to u64. Afaict += should be equivalent to |= if the only active bytes are above the window this is casting bytes in to.

Reply:
It works, problem of the previous implementation in case you have e.g. -1(i8) and try to cast it to i64, the result will be `0b1111111111111111111111111111111111111111111111111111111111111111` and again for instance you have `self_width == 8` and `shift == 48` - you'll get `0b1111111111111111000000000000000000000000000000000000`, but you need to have `0b00000000011111111000000000000000000000000000000000000` so as not to affect the bits of the other part of the encoding.
If I cast to i64 right away, the right shift will change the most significant bits to 1 instead of 0.

Yesterday I had a lot of interesting insights :D

Reply:
Oh i see. Is this also achieved by casting to the unsigned equivalent before u64? That’s what we do elsewhere afaik.

e.g. i16 as u16 as u64, do the shift, as i64

Reply:
I think that two casts in a row is absolutely useless in all cases. But maybe I don't know something.

Reply:
It’s counterintuitive, but casting from i16 -> u16 -> u64 -> i64 is different from i16 -> i64 because of how the signed/unsigned casts treat negative values